### PR TITLE
Update command guidelines

### PR DESCRIPTION
**Pull request by derekbekoe**

### CODE REVIEW

**Comment by promptws** [comment]
View a preview at https://prompt.ws/r/Azure/azure-cli/5769
This is an **experimental preview** for @microsoft users.

File: doc/command_guidelines.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+- `LIST` - command to list instances of a resource, backed server-side by a GET request. When there are multiple "list-type" commands within an SDK to list resources at different levels (for example, listing resources in a subscription vice in a resource group) the functionality should be exposed by have a single list command with arguments to control the behavior. For example, if `--resource-group` is provided, the command will call `list_by_resource_group`; otherwise, it will call `list_by_subscription`.
+- `DELETE` - command to delete a resource, backed server-side by a DELETE request. Delete commands return nothing on success.
+
+## Non-standard Commands
+
+For commands that don't conform to one of the above-listed standard command patterns, use the following guidance.
+
+- (*) Don't use single word verbs if they could cause confusion with the standard command types. For example, don't use `get` or `new` as these sound functionally the same as `show` and `create` respectively, leading to confusion as to what the expected behavior should be.
+e.g. `new`, `get`
+- (*) Descriptive, hyphenated command names are often a better option than single verbs.
```
Comment (tjprescott) [inline doc/command_guidelines.md]: The use of `(*)` was a confusing to me at first--I thought it was mangled markup. Then I scrolled back to the top and saw.  You might simply call this out in-text:
"Blah blah blah, don't use new *(modules only)*"

File: doc/command_guidelines.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+- `SHOW` - command to show the properties of a resource, backed server-side by a GET request.
+- `LIST` - command to list instances of a resource, backed server-side by a GET request. When there are multiple "list-type" commands within an SDK to list resources at different levels (for example, listing resources in a subscription vice in a resource group) the functionality should be exposed by have a single list command with arguments to control the behavior. For example, if `--resource-group` is provided, the command will call `list_by_resource_group`; otherwise, it will call `list_by_subscription`.
+- `DELETE` - command to delete a resource, backed server-side by a DELETE request. Delete commands return nothing on success.
+
+## Non-standard Commands
+
+For commands that don't conform to one of the above-listed standard command patterns, use the following guidance.
+
+- (*) Don't use single word verbs if they could cause confusion with the standard command types. For example, don't use `get` or `new` as these sound functionally the same as `show` and `create` respectively, leading to confusion as to what the expected behavior should be.
+e.g. `new`, `get`
```
Comment (tjprescott) [inline doc/command_guidelines.md]: Why is this line here?  It's already mentioned in the text.

File: doc/command_guidelines.md [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+- Multi-word subgroups should be hyphenated
+e.g. `foo-resource` instead of `fooresource`
+- All command names should contain a verb
+e.g. `account get-connection-string` instead of `account connection-string`
+- Avoid hyphenated command names when moving the commands into a subgroup would eliminate the need.
+e.g. `database show` and `database get` instead of `show-database` and `get-database`
+- If a command subgroup would only have a single command, move it into the parent command group and hyphenate the name. This is common for commands which exist only to pull down cataloging information.
+e.g. `database list-sku-definitions` instead of `database sku-definitions list`
+- Avoid command subgroups that have no commands. This often happens at the first level of a command branch.
+e.g. `keyvault create` instead of `keyvault vault create` (where `keyvault` only has subgroups and adds unnecessary depth to the tree).
```
Comment (tjprescott) [inline doc/command_guidelines.md]: I prefer the original, even though it is longer. The example would likely not be helpful at all to someone who isn't us without the preserved help output. When I read this, it sounds like it's keyvault specific and misses the bigger picture.

File: doc/command_guidelines.md [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+- Support tab completion for parameter names and values (e.g. resource names)
+- Commands must follow a "[noun] [noun] [verb]" pattern
+- For nouns that only support a single verb, the command should be named as a single hyphenated verb-noun pair
+- All commands, command group and arguments must have descriptions
+- You must provide command examples for non-trivial commands
+- All commands must support all output types (json, tsv, table)
+- Provide custom table outputs for commands that don't provide table output automatically
+- Commands must return an object or dictionary (do not string, Boolean, etc. types)
+- Use `stdout` and `stderr` appropriately.
+- Command output must go to stdout, everything else to stderr (log/status/errors).
```
Comment (tjprescott) [inline doc/command_guidelines.md]: Since this line defines what "appropriately" is for stdout and stderr, does line 20 need to be said?

File: doc/command_guidelines.md [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+- Be consistent with POSIX tools (support piping, work with grep, awk, jq, etc.)
+- Support tab completion for parameter names and values (e.g. resource names)
+- Commands must follow a "[noun] [noun] [verb]" pattern
+- For nouns that only support a single verb, the command should be named as a single hyphenated verb-noun pair
+- All commands, command group and arguments must have descriptions
+- You must provide command examples for non-trivial commands
+- All commands must support all output types (json, tsv, table)
+- Provide custom table outputs for commands that don't provide table output automatically
+- Commands must return an object or dictionary (do not string, Boolean, etc. types)
```
Comment (tjprescott) [inline doc/command_guidelines.md]: This makes it sounds like all commands must return something, which isn't the case.

File: doc/command_guidelines.md [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+## General Patterns
+
+- Be consistent with POSIX tools (support piping, work with grep, awk, jq, etc.)
+- Support tab completion for parameter names and values (e.g. resource names)
+- Commands must follow a "[noun] [noun] [verb]" pattern
+- For nouns that only support a single verb, the command should be named as a single hyphenated verb-noun pair
+- All commands, command group and arguments must have descriptions
+- You must provide command examples for non-trivial commands
+- All commands must support all output types (json, tsv, table)
```
Comment (tjprescott) [inline doc/command_guidelines.md]: This isn't strictly true.  There are some convenience commands that output in JSON only regardless of format, or the ACS command that outputs YAML.

File: doc/command_guidelines.md [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+Guidelines marked (*) only apply to command modules, not extensions.
+
+If in doubt, ask!
+
+## General Patterns
+
+- Be consistent with POSIX tools (support piping, work with grep, awk, jq, etc.)
+- Support tab completion for parameter names and values (e.g. resource names)
+- Commands must follow a "[noun] [noun] [verb]" pattern
+- For nouns that only support a single verb, the command should be named as a single hyphenated verb-noun pair
```
Comment (tjprescott) [inline doc/command_guidelines.md]: This is stated later on.

File: doc/command_guidelines.md [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+Guidelines marked (*) only apply to command modules, not extensions.
+
+If in doubt, ask!
+
+## General Patterns
+
+- Be consistent with POSIX tools (support piping, work with grep, awk, jq, etc.)
+- Support tab completion for parameter names and values (e.g. resource names)
+- Commands must follow a "[noun] [noun] [verb]" pattern
```
Comment (tjprescott) [inline doc/command_guidelines.md]: Consider moving this down into the "Command Naming and Behavior Guidance" section.

File: doc/command_guidelines.md [hunk 9]
```diff
@@ -0,0 +1,63 @@
+# Command Guidelines
+
+This document describes the command guidelines for 'az' and applies to both CLI command modules and extensions.
+
+Guidelines marked (*) only apply to command modules, not extensions.
+
+If in doubt, ask!
+
+## General Patterns
```
Comment (tjprescott) [inline doc/command_guidelines.md]: It seems like some of the stuff in General Patterns would better be placed under other subheadings. I call out a few but best to consider them all.

File: doc/command_guidelines.md (hunk 2 above)
Comment (derekbekoe) [reply doc/command_guidelines.md]: removed.

File: doc/command_guidelines.md (hunk 4 above)
Comment (derekbekoe) [reply doc/command_guidelines.md]: removed that line.

File: doc/command_guidelines.md (hunk 5 above)
Comment (derekbekoe) [reply doc/command_guidelines.md]: I've clarified that `None` can be returned also.

File: doc/command_guidelines.md (hunk 6 above)
Comment (derekbekoe) [reply doc/command_guidelines.md]: Those are exceptions that aren't ideal so would rather not call that out as part of the guidelines doc which should be recommendations on what you *should* be doing.

File: doc/command_guidelines.md (hunk 7 above)
Comment (derekbekoe) [reply doc/command_guidelines.md]: removed

File: doc/command_guidelines.md (hunk 8 above)
Comment (derekbekoe) [reply doc/command_guidelines.md]: done

File: doc/command_guidelines.md (hunk 9 above)
Comment (derekbekoe) [reply doc/command_guidelines.md]: I think a separation between 'general patterns' and other things that pertain only to 'command naming' etc. makes sense.
Arguably everything could be put under 'Behavior' but keeping each section a few points each makes things clearer I think.

File: doc/command_guidelines.md (hunk 1 above)
Comment (derekbekoe) [reply doc/command_guidelines.md]: I do call it out at the top so now that you know what the `(*)` means, does it make sense?
I didn't want to prefix each of those lines with `(command modules only)` as it's kinda verbose.

Comment (tjprescott) [reply doc/command_guidelines.md]: Assuming everyone reads the entire document from the top, and pays attention, then yes.  Many people I think will simply scroll to where ever they think the info they want is, especially if they follow a link to a particular section.  Since it's a help document, being verbose is probably better than being cryptic, but I am fine if you want to leave it as is.

File: doc/command_guidelines.md (hunk 3 above)
Comment (derekbekoe) [reply doc/command_guidelines.md]: added to an expandable section.

File: doc/command_guidelines.md (hunk 6 above)
Comment (tjprescott) [reply doc/command_guidelines.md]: Agreed. Perhaps then change "must" to "should" :)
